core/src/main/resources/META-INF/native-image/com.datastax.oss/java-driver-core/reflection.json

```diff
@@ -0,0 +1,214 @@
+  {
+    "name": "com.datastax.oss.driver.internal.core.retry.DefaultRetryPolicy",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.specex.NoSpeculativeExecutionPolicy",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.metadata.NoopNodeStateListener",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.tracker.NoopRequestTracker",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.session.throttling.PassThroughRequestThrottler",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.time.AtomicTimestampGenerator",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.channel.InFlightHandler",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.channel.ProtocolInitHandler",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.protocol.FrameDecoder",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.protocol.FrameEncoder",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.channel.ChannelFactory$1",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.channel.HeartbeatHandler",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
+  {
+    "name": "com.datastax.oss.driver.internal.core.metrics.DefaultMetricsFactory",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allPublicMethods": true
+  },
```

Comment:
Entries above largely reflect the driver defaults as of 4.11.  Note that this also includes the Dropwizard -> Default MetricsFactory change introduced with JAVA-2927.

Reply:
I analyzed briefly the reference.conf file for 4.12.0 and found the following interfaces and implementing classes:

```
com.datastax.oss.driver.api.core.loadbalancing.LoadBalancingPolicy
    * com.datastax.oss.driver.internal.core.loadbalancing.DefaultLoadBalancingPolicy
    com.datastax.oss.driver.internal.core.loadbalancing.DcInferringLoadBalancingPolicy
    com.datastax.oss.driver.internal.core.loadbalancing.BasicLoadBalancingPolicy

com.datastax.oss.driver.api.core.connection.ReconnectionPolicy
    * com.datastax.oss.driver.internal.core.connection.ExponentialReconnectionPolicy
    com.datastax.oss.driver.internal.core.connection.ConstantReconnectionPolicy

com.datastax.oss.driver.api.core.retry.RetryPolicy
    * com.datastax.oss.driver.internal.core.retry.DefaultRetryPolicy
    com.datastax.oss.driver.internal.core.retry.ConsistencyDowngradingRetryPolicy

com.datastax.oss.driver.api.core.retry.RetryPolicy
    * com.datastax.oss.driver.internal.core.retry.DefaultRetryPolicy
    com.datastax.oss.driver.internal.core.retry.ConsistencyDowngradingRetryPolicy

com.datastax.oss.driver.api.core.specex.SpeculativeExecutionPolicy
    * com.datastax.oss.driver.internal.core.specex.NoSpeculativeExecutionPolicy
    com.datastax.oss.driver.internal.core.specex.ConstantSpeculativeExecutionPolicy

com.datastax.oss.driver.api.core.time.TimestampGenerator
    * com.datastax.oss.driver.internal.core.time.AtomicTimestampGenerator
    com.datastax.oss.driver.internal.core.time.ThreadLocalTimestampGenerator
    com.datastax.oss.driver.internal.core.time.ServerSideTimestampGenerator

com.datastax.oss.driver.api.core.tracker.RequestTracker
    * com.datastax.oss.driver.internal.core.tracker.NoopRequestTracker
    com.datastax.oss.driver.internal.core.tracker.RequestLogger
    com.datastax.dse.driver.internal.core.tracker.MultiplexingRequestTracker

com.datastax.oss.driver.api.core.session.throttling.RequestThrottler
    * com.datastax.oss.driver.internal.core.session.throttling.PassThroughRequestThrottler
    com.datastax.oss.driver.internal.core.session.throttling.RateLimitingRequestThrottler
    com.datastax.oss.driver.internal.core.session.throttling.ConcurrencyLimitingRequestThrottler

com.datastax.oss.driver.api.core.metadata.NodeStateListener
    * com.datastax.oss.driver.internal.core.metadata.NoopNodeStateListener
    com.datastax.oss.driver.api.core.metadata.SafeInitNodeStateListener

com.datastax.oss.driver.api.core.metadata.schema.SchemaChangeListener
    * com.datastax.oss.driver.internal.core.metadata.schema.NoopSchemaChangeListener

com.datastax.oss.driver.api.core.addresstranslation.AddressTranslator
    * com.datastax.oss.driver.internal.core.addresstranslation.PassThroughAddressTranslator
    com.datastax.oss.driver.internal.core.addresstranslation.Ec2MultiRegionAddressTranslator

com.datastax.oss.driver.api.core.auth.AuthProvider
    com.datastax.oss.driver.internal.core.auth.PlainTextAuthProvider
    com.datastax.oss.driver.internal.core.auth.ProgrammaticPlainTextAuthProvider
    com.datastax.dse.driver.internal.core.auth.DseGssApiAuthProvider
    com.datastax.dse.driver.api.core.auth.ProgrammaticDseGssApiAuthProvider

com.datastax.oss.driver.api.core.ssl.SslEngineFactory
    com.datastax.oss.driver.internal.core.ssl.DefaultSslEngineFactory
    com.datastax.oss.driver.api.core.ssl.ProgrammaticSslEngineFactory
    com.datastax.oss.driver.internal.core.ssl.SniSslEngineFactory

com.datastax.oss.driver.internal.core.metrics.MetricsFactory
    * com.datastax.oss.driver.internal.core.metrics.DefaultMetricsFactory
    com.datastax.oss.driver.internal.core.metrics.NoopMetricsFactory
    com.datastax.oss.driver.internal.core.metrics.DropwizardMetricsFactory
    com.datastax.oss.driver.internal.metrics.microprofile.MicroProfileMetricsFactory
    com.datastax.oss.driver.internal.metrics.micrometer.MicrometerMetricsFactory

com.datastax.oss.driver.internal.core.metrics.MetricIdGenerator
    * com.datastax.oss.driver.internal.core.metrics.DefaultMetricIdGenerator
    com.datastax.oss.driver.internal.core.metrics.TaggingMetricIdGenerator
```

Default implementations are marked with an asterisk.

This list brings the following remarks:

1. You are currently including only the default impls. We have discussed this already offline but haven't came to an agreement: should we include all impls? I understand the risks (larger image, warnings during image build).
2. A few interfaces are missing:
  a. `AuthProvider`
  b. `SslEngineFactory`
  c. `MetricIdGenerator` (introduced recently by JAVA-2872 in 4.11.0)
3. I am not sure about these ones; they are all handlers instantiated inside a `ChannelInitializer` inside `ChannelFactory` (the first one being the anonymous `ChannelInitializer` itself) :
  a. `com.datastax.oss.driver.internal.core.channel.ChannelFactory$1`
  b. `com.datastax.oss.driver.internal.core.channel.InFlightHandler`
  c. `com.datastax.oss.driver.internal.core.channel.ProtocolInitHandler`
  d. `com.datastax.oss.driver.internal.core.protocol.FrameDecoder`
  e. `com.datastax.oss.driver.internal.core.protocol.FrameEncoder`
  f. `com.datastax.oss.driver.internal.core.channel.HeartbeatHandler`
4. I am also not sure about the Netty ones below, and especially worried about these getting out of sync when Netty is upgraded:
  a. `io.netty.channel.socket.nio.NioSocketChannel`
  b. `io.netty.buffer.AbstractByteBufAllocator`
  c. `io.netty.util.ReferenceCountUtil`
  d. `io.netty.channel.DefaultChannelPipeline$HeadContext`
  e. `io.netty.channel.DefaultChannelPipeline$TailContext`
  f. `io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueProducerIndexField`
  g. `io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueProducerLimitField`
  h. `io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueueConsumerIndexField`
  i. `io.netty.util.internal.shaded.org.jctools.queues.BaseMpscLinkedArrayQueueProducerFields`
  j. `io.netty.util.internal.shaded.org.jctools.queues.BaseMpscLinkedArrayQueueColdProducerFields`
  k. `io.netty.util.internal.shaded.org.jctools.queues.BaseMpscLinkedArrayQueueConsumerFields`

Reply:
Nit: `com.datastax.oss.driver.internal.core.channel.ChannelFactory$1` is a generated anonymous class name that could also get out of sync if we introduce another anonymous class in `ChannelFactory`. It might be a good idea to extract that anonymous class to an inner.

Reply:
I'm still working through the contents of this file but a few notes as I go here.

* NioSocketChannel has been removed.  Originally I thought we had to keep it to support the Helidon case but it looks very much like the problem there is a Helidon issue.  More detail on this at https://github.com/datastax/java-driver/pull/1560/commits/ea5258065cb996b6f92cd222c0c7478a54c89213
* The shaded jctools stuff certainly appears to be required, at least for the version of Netty we're using.  See comments below for the relevant Netty issue.  I agree that maintaining this kind of thing is a pain but I'm not sure what other option we have.
* I'm still testing/spelunking the other Netty classes to see what else (if anything) can be removed from this list
* I very much like the idea of abstracting the ChannelFactory anonymous class into an inner class... I'll try to fold that into other work going in soon.

Reply:
Update: all the Netty classes that aren't associated with the shaded jctools have been removed based on local testing

Reply:
Update: NioSocketChannel and friends are back, although in stripped down form.  Subsequent testing identified an unexpected reflection config which was in play during earlier testing; that config was effectively masking a trio of Netty-related issues from driver code.  These have now been resolved.

Reply:
Update: re: the classes from the channel initializer... I'm not sure why these are present either.  I would expect static analysis to come across these pretty readily.  Testing with the current working version of this PR has confirmed that the removal of these entries in a reflection config doesn't have any adverse effects on our test cases.  I suppose this _could be_ due to the refactoring of the ChannelFactory above but I really don't see how... I'd expect similar analysis for both the before and after code.

Reply:
Only thing really left here were the three potential new candidates for inclusion: impls for AuthProvider, SslEngineFactory and MetricIdGenerator.  These seem a bit more complicated to me; they all come into play for optional features of the driver but aren't used in the base no-auth/no-ssl case.  I could perhaps see different handling for MetricIdGenerator since we at least have a default defined for that in reference.conf... but that isn't true for AuthProvider and SslEngineFactory.

We already know that we'll have to provide guidance for users who wish to use non-standard components + Graal.  I'm wondering if we shouldn't take a similar line for these three components as well; if you want to use them you have to do some additional manual customizations, specifically create your own local reflection.json file to supplement the one the driver provides (via this PR).

@adutra whaddya think?

Reply:
Thanks for the latest changes regarding ChannelFactory and Netty, these areas are looking much better now.

Now back to the question of what classes to include here, only defaults vs all driver built-ins. I am still struggling with 2 fundamental questions:

1. Who is this file going to be useful for?
2. What contents would make it the most useful for most people?

For 1), I can think of two personas: end users, and maintainers of microservice frameworks. Let's see now how to get at 2) for these two cases:

* **As an end user building my native image myself.** 
    * Suppose that only default impls are declared here. If I bootstrap my application using only driver defaults, the native image would build without issues thanks to this file. However if I later on decide to use, say, `RateLimitingRequestThrottler` instead of `PassThroughRequestThrottler` by declaring this in an application.conf file, or using the programmatic config builder, then all of a sudden, the native image build will break. This isn't great in terms of usability because the config change seems unrelated to the build failure. Without proper guidance/documentation, I bet that user would struggle to fix this. (That said, I admit that this user would also need to register their application.conf file if they are using one... I don't think we can do this for them).
    * Suppose now that all built-in impls are declared here. The above config change wouldn't break the native image build (provided that application.conf is registered if used). 
    * I therefore think that for this persona, having all built-in impls declared here is better.
    * Obviously, if that user later on changes `RateLimitingRequestThrottler` for a custom impl of `RequestThrottler`, the native image build will fail again if the custom impl is not registered. I would argue that such a user is a more advanced one, and by the time they start implementing driver interfaces, they should be aware of the implications.
    * Also obviously, the more you declare for reflection, the bigger your native image will be. I think we are trading complexity against disk space here. I'd argue that disk space is cheaper :-)
   
* **As a microservices framework maintainer (Quarkus, Micronaut, Helidon)**. 
    * If only default impls are declared here, but if the framework allows some of these defaults to be replaced (e.g. Quarkus allows the metrics factory to be replaced), then this file is kind of pointless, as the framework will have to provide at least a complement for its contents, if not a complete replacement. This is precisely what Quarkus is doing already, e.g for [metrics](https://github.com/datastax/cassandra-quarkus/blob/main/deployment/src/main/java/com/datastax/oss/quarkus/deployment/internal/CassandraClientProcessor.java#L247).
    * Suppose now that all built-in impls are declared here. A config change such as setting the metrics factory to Micrometer for instance, made through the framework config mechanism or through a driver-specific one, wouldn't break the native image build. 
    * So again, I think that for this persona, having all built-in impls declared here is better. The framework maintainers in question wouldn't need to provide any additional Graal customization – unless they want to because they want a more fine-grained control over the native image contents (like we do in Quarkus for metrics – we'll probably keep doing it even after this change is merged).
    * Such microservices frameworks generally discourage users from bypassing their config mechanisms when configuring the driver. E.g. the usage of an application.conf file in Quarkus is possible, although discouraged. If such a file is included in a Quarkus app, the user would need to register it manually. And if such a file contains a custom impl of a driver interface, that would have to be registered too. Again, we are here facing a rather advanced user trying to go beyond what the framework is able to provide. I think we can safely consider this guy as a user that know what he's doing.

So to summarize my POV: including everything we can here would satisfy a broader range of users and require less work from them. This seems especially true for Micronaut and Helidon users.

Reply:
I'm sympathetic to the position you lay out @adutra, but I think I still come down on the other side.  I'll stick with your framework because I think it's a useful way of formulating the questions.

My recommended approach is the following:

* Provide a minimal config embedded in the JAR that only includes the default instances in a reflection config
* Any users wishing to include custom component impls (or other config customizations) must supply additional configs to support such work.  We would supply docs to communicate this and provide some guidance on how to do so.

My rationale runs as follows:

* Native image size is a factor for Graal users; I've seen it mentioned as a concern on any number of tickets/conversations around the ecosystem
  * If we include all the classes in our reflection config there's no way for users to _not_ include those classes in the image; subsequent config files can't unselect a class which was previously selected for inclusion.  The only way to undo it would be to edit the config directly and update the config in the JAR.
* The number of "end users" (again using the language of your scenarios) is fairly small.  The intersection of those users with users who wish to leverage custom driver components is either empty or very nearly empty (at least for the moment).  As a result the default configs allow us to address all or nearly all "end user" cases without including the additional classes (and taking on the additional image size)
* Shifting to frameworks, these would have to take additional steps to properly configure the drivers if a users selects a custom feature at a framework level.  As you point out we do this in the Quarkus plugin, but I'd argue that's exactly the right place to do it.  
  * Frameworks already have built-in mechanisms for including additional resources in the native image build
  * There doesn't seem to be a clear argument for _also_ implementing a similar feature at the driver level itself, especially if we grant the argument (outlined above) that the population of "end users" who wish to leverage such a similar feature at a driver level is vanishingly small.
* In the medium-to-longer term we can more thoroughly investigate the idea of adding a feature which will perform the selective loading of dependencies at a driver level
  * But I would like to see additional feedback from the "end users" in your scenario saying this is useful before we add the additional complexity here

In summary, I guess I'm thinking of the following:

* "Always add" the smallest amount we can to maximize functionality while keeping the image size small
* Work with framework authors to make sure the framework case can leverage the infrastructure it already has to do the right job
* Revise our plan if we get feedback from actual users saying our assumptions above are flawed

Reply:
I'm also very sympathetic to your position @absurdfarce :-) And I could easily let myself convince to adopt your suggestions.

But in this case I think we should go even further: why even provide a minimal reflection config? 

* As you rightly pointed out, "native image size is a factor for Graal users". Which means that even the minimal config could add some clutter to the native image. E.g. if the request throttler is changed, the default one would still be registered for reflection, and occupy some space for nothing.
* "[Frameworks] would have to take additional steps to properly configure the drivers if a users selects a custom feature at a framework level": yes. And that's the core of the question. If we go down your route, how is this minimal config going to be useful to, say, Iván López and his struggle to make Micronaut work seamlessly with the Java driver? He would still have to implement for Micronaut a fair amount of the logic that we put in the Quarkus extension. The goal of this ticket was to give Iván a clean base: without any further customization on his side, the driver would still be guaranteed to work and generate a valid native image – even if not the most optimized one.
* And let's not forget the case where a user of a framework is bypassing its configuration mechanism and configuring the driver directly. This is the case, for instance, of all DSE users using Quarkus currently. How is the framework going to detect the bits of configuration being done outside of its realm? If it cannot detect those bits, how is reflection going to be configured, by the end users themselves?

In summary, I can totally see 2 valid options: register for reflection everything that ships with the driver, or nothing at all. But registering just the defaults does seem to provide enough added-value to justify the inclusion fo these configs in the driver jar.

The "nothing at all" option is actually very appealing. If for nothing else, it has the advantage of letting the user decide what they need to include. This is clearly better than the "include all" option at least under one aspect: once some class is registered for reflection, there is no way to "un-register" it. 

Why don't we go down that route then? In this case the work to do here could be restricted to just:

* Declaring a few resources (reference.conf, Driver.properties);
* Declaring the TypeVariable proxy;
* And documenting the reflection side of things: how to properly declare classes for reflection depending on what is going to be used. But no class would be registered by default.
* Additionally, the documentation could include a giant "reflection.json" config file containing all the impls that ship with the driver. That file could serve as a starting point for users trying to create their own file.

Wdyt?

Reply:
For future reference: after feedback from the Micronaut team, we decided to go with the option of providing automatic registration for reflection of all driver built-in impls.